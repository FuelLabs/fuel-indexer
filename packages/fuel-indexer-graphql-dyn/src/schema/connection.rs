use super::node::*;
use super::paging::*;
use super::self_prelude::*;

#[derive(Clone, PartialEq, Eq)]
pub struct DynConnectionEdge {
    pub node_id: DynNodeId,
    pub cursor: Cursor,
}

#[derive(Clone, PartialEq, Eq)]
pub struct DynConnection {
    pub total_count: usize,
    pub edges: Vec<DynConnectionEdge>,
    pub page_info: DynPageInfo,
}

impl DynConnectionEdge {
    pub fn new(
        node_id: &(impl Into<DynNodeId> + Clone),
        cursor: &(impl Into<Cursor> + Clone),
    ) -> Self {
        Self {
            node_id: node_id.clone().into(),
            cursor: cursor.clone().into(),
        }
    }

    pub fn node_id(&self) -> &DynNodeId {
        &self.node_id
    }
    pub fn cursor(&self) -> &Cursor {
        &self.cursor
    }
}
