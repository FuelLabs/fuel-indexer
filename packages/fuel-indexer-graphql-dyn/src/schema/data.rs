use super::self_prelude::*;

#[derive(Clone, Debug, Hash, PartialEq, Eq, Serialize, Deserialize)]
pub enum DynDataTypeId {
    // GraphQL types
    ID,
    String,
    Int,
    Float,
    Boolean,
    List(Box<DynDataTypeId>),
    // Option(Box<DynDataTypeId>),
    // Fuel types
    Unit,
    U8,
    U16,
    U32,
    U64,
    B256,
    Bytes,
    Name(Name),
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum DynDataType {
    // GraphQL types
    ID,
    String,
    Int,
    Float,
    Boolean,
    List(Box<DynDataType>),
    // Option(Box<DynDataTypeId>),
    // Fuel types
    Unit,
    U8,
    U16,
    U32,
    U64,
    B256,
    Bytes,
    Object(Name, IndexMap<DynDataFieldId, DynDataFieldType>),
    Enum(Name, IndexMap<DynDataFieldId, DynDataFieldType>),
}

pub type DynDataFieldId = String;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DynDataFieldType {
    pub name: Name,
    pub data_type_id: DynDataTypeId,
    pub store_data_field_id: store::DataFieldId,
}

impl From<store::DataTypeId> for DynDataTypeId {
    fn from(store_data_type_id: store::DataTypeId) -> Self {
        match store_data_type_id {
            store::DataTypeId::Unit => Self::Unit,
            store::DataTypeId::Bool => Self::Boolean,
            store::DataTypeId::U8 => Self::U8,
            store::DataTypeId::U16 => Self::U16,
            store::DataTypeId::U32 => Self::U32,
            store::DataTypeId::U64 => Self::U64,
            store::DataTypeId::B256 => Self::B256,
            store::DataTypeId::Byte => Self::U8,
            store::DataTypeId::Bytes => Self::Bytes,
            store::DataTypeId::String => Self::String,
            store::DataTypeId::Array(ty) => Self::List(Box::new((*ty).into())),
            store::DataTypeId::Name(name) => Self::Name(name),
        }
    }
}

impl FromStr for DynDataTypeId {
    type Err = ();
    fn from_str(str: &str) -> Result<Self, Self::Err> {
        match str {
            // GraphQL types
            "ID" => Ok(Self::ID),
            "String" => Ok(Self::String),
            "Int" => Ok(Self::Int),
            "Float" => Ok(Self::Float),
            "Boolean" => Ok(Self::Boolean),
            // List(Box<DynDataTypeId>),
            // Option(Box<DynDataTypeId>),
            // Fuel types
            "Unit" => Ok(Self::Unit),
            "U8" => Ok(Self::U8),
            "U16" => Ok(Self::U16),
            "U32" => Ok(Self::U32),
            "U64" => Ok(Self::U64),
            "B256" => Ok(Self::B256),
            "Bytes" => Ok(Self::Bytes),
            str => Ok(Self::Name(Name::new_pascal(str))),
        }
    }
}

impl std::fmt::Display for DynDataTypeId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::ID => write!(f, "ID"),
            Self::Int => write!(f, "Int"),
            Self::Float => write!(f, "Float"),
            Self::List(ty) => write!(f, "[{}]", ty),
            // Self::Option(ty) => write!(f, "{}", ty),
            Self::Unit => write!(f, "Unit"),
            Self::Boolean => write!(f, "Boolean"),
            Self::U8 => write!(f, "U8"),
            Self::U16 => write!(f, "U16"),
            Self::U32 => write!(f, "U32"),
            Self::U64 => write!(f, "U64"),
            Self::B256 => write!(f, "B256"),
            Self::Bytes => write!(f, "Bytes"),
            Self::String => write!(f, "String"),
            Self::Name(str) => write!(f, "{}", str),
        }
    }
}

impl DynDataType {
    pub fn name(&self) -> Name {
        match self {
            Self::ID => Name::new_pascal("ID"),
            Self::Int => Name::new_pascal("Int"),
            Self::Float => Name::new_pascal("Float"),
            Self::List(ty) => Name::new_pascal(format!("[{}]", ty.name())),
            // Self::Option(ty) => Name::new_pascal(&format!("{}!", ty)),
            Self::Unit => Name::new_pascal("Unit"),
            Self::Boolean => Name::new_pascal("Boolean"),
            Self::U8 => Name::new_pascal("U8"),
            Self::U16 => Name::new_pascal("U16"),
            Self::U32 => Name::new_pascal("U32"),
            Self::U64 => Name::new_pascal("U64"),
            Self::B256 => Name::new_pascal("B256"),
            Self::Bytes => Name::new_pascal("Bytes"),
            Self::String => Name::new_pascal("String"),
            Self::Object(name, _) => Name::new_pascal(format!("{}", name)),
            Self::Enum(name, _) => Name::new_pascal(format!("{}", name)),
        }
    }
}

impl DynDataFieldType {
    pub fn data_type_id(&self) -> &DynDataTypeId {
        &self.data_type_id
    }
    pub fn store_id(&self) -> &store::DataFieldId {
        &self.store_data_field_id
    }
}
