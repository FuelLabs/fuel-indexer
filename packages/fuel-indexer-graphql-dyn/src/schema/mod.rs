pub mod connection;
pub mod data;
pub mod edge;
pub mod loader;
pub mod node;
pub mod paging;
pub mod resolver;
pub mod resolver_context;
pub mod schema_builder;
pub mod schema_type;
pub mod schema_type_builder;

pub(self) mod self_prelude {
    pub use super::super::self_prelude::*;
    pub use crate::spec::*;
    pub use crate::store;
    pub use crate::store::Name;
    pub use async_graphql::dynamic::*;
    pub use indexmap::IndexMap;
    pub use serde::{Deserialize, Serialize};
    pub use std::collections::HashMap;
    pub use std::fmt;
    pub use std::str::FromStr;
    pub use std::sync::Arc;
    pub use tokio::sync::Mutex;
}

pub use connection::*;
pub use data::*;
pub use edge::*;
pub use loader::*;
pub use node::*;
pub use paging::*;
pub use resolver::*;
pub use resolver_context::*;
pub use schema_builder::*;
pub use schema_type::*;
pub use schema_type_builder::*;
