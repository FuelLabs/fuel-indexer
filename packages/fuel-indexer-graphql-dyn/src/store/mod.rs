pub mod assoc;
pub mod assoc_type;
pub mod data;
pub mod data_type;
pub mod meta;
pub mod obj;
pub mod obj_type;
#[allow(clippy::module_inception)]
pub mod store;
pub mod store_type;
pub mod store_type_builder;

pub(self) mod self_prelude {
    pub use super::super::self_prelude::*;
    pub use indexmap::IndexMap;
    pub use serde::{Deserialize, Serialize};
    pub use std::fmt;
    pub use std::str::FromStr;
}

pub use assoc::*;
pub use assoc_type::*;
pub use data::*;
pub use data_type::*;
pub use meta::*;
pub use obj::*;
pub use obj_type::*;
pub use store::*;
pub use store_type::*;
pub use store_type_builder::*;
