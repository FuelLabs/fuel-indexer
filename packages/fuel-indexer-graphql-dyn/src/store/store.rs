use super::assoc::*;
use super::obj::*;
use super::self_prelude::*;
use super::store_type::*;

type Result<T> = anyhow::Result<T, StoreError>;

/// See: https://research.facebook.com/publications/tao-facebooks-distributed-data-store-for-the-social-graph/
#[async_trait]
pub trait Store: Send + Sync {
    /// Returns the type of the store.
    fn r#type(&self) -> &StoreType;
    //
    /// Gets a single object by its ID.
    async fn obj_get(&self, id: &ObjId) -> Result<Option<Obj>>;
    /// Gets objects by their IDs.
    async fn obj_get_many(&self, ids: &[ObjId]) -> Result<Vec<Option<Obj>>>;
    // Gets associations of an object by their head IDs.
    async fn assoc_get(
        &self,
        key: &AssocKey,
        ids: &[ObjId],
        high: Option<AssocTime>,
        low: Option<AssocTime>,
    ) -> Result<Vec<Option<Assoc>>>;
    // Gets the count of associations an object has.
    async fn assoc_count(&self, key: &AssocKey) -> Result<u64>;
    // Gets associations of an object.
    async fn assoc_range(
        &self,
        key: &AssocKey,
        offset: u64,
        limit: u64,
    ) -> Result<Vec<Assoc>>;
    // Gets associations of an object in the given time range.
    async fn assoc_time_range(
        &self,
        key: &AssocKey,
        high: AssocTime,
        low: AssocTime,
        limit: u64,
    ) -> Result<Vec<Assoc>>;
}

pub type StoreResult<T> = anyhow::Result<T, StoreError>;

#[derive(thiserror::Error, Debug)]
pub enum StoreError {
    #[error(transparent)]
    Other(#[from] anyhow::Error),
}
