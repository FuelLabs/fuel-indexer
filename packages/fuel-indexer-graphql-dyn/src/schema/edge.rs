use super::data::*;
use super::node::*;
use super::self_prelude::*;

pub type DynEdgeTypeId = store::AssocTypeId;
pub type DynEdgeData = store::Assoc;
pub type DynEdgeTime = store::AssocTime;

#[derive(Clone, Hash, PartialEq, Eq)]
pub struct DynEdgeId(DynEdgeTypeId, DynNodeId, DynNodeId);

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DynEdgeType {
    pub name: Name,
    pub tail: DynNodeTypeId,
    pub head: DynNodeTypeId,
    pub fields: IndexMap<DynDataFieldId, DynDataFieldType>,
}

#[derive(Clone)]
pub struct DynEdge(DynEdgeTypeId, DynNodeId, DynEdgeData);

impl DynEdgeId {
    pub fn new(
        type_id: impl Into<DynEdgeTypeId>,
        tail_id: impl Into<DynNodeId>,
        head_id: impl Into<DynNodeId>,
    ) -> Self {
        Self(type_id.into(), tail_id.into(), head_id.into())
    }
}

impl DynEdgeType {
    pub fn tail_type_id(&self) -> &DynNodeTypeId {
        &self.tail
    }
    pub fn head_type_id(&self) -> &DynNodeTypeId {
        &self.head
    }
    pub fn fields(&self) -> &IndexMap<DynDataFieldId, DynDataFieldType> {
        &self.fields
    }
    pub fn field(&self, field_id: &DynDataFieldId) -> Option<&DynDataFieldType> {
        self.fields.get(field_id)
    }
}

impl DynEdge {
    pub fn new(
        type_id: impl Into<DynEdgeTypeId>,
        tail_id: impl Into<DynNodeId>,
        data: impl Into<DynEdgeData>,
    ) -> Self {
        Self(type_id.into(), tail_id.into(), data.into())
    }

    pub fn type_id(&self) -> &DynEdgeTypeId {
        &self.0
    }
    pub fn tail_id(&self) -> &DynNodeId {
        &self.1
    }
    pub fn head_id(&self) -> &DynNodeId {
        self.2.id()
    }
    pub fn time(&self) -> &DynEdgeTime {
        self.2.time()
    }
    pub fn data(&self) -> &serde_json::Value {
        self.2.data().value()
    }
}
