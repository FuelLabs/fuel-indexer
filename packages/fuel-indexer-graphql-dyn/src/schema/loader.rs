use super::edge::*;
use super::node::*;
use super::self_prelude::*;

pub struct DynLoader {
    pub store: Arc<Mutex<dyn store::Store>>,
}

pub type DynLoaderResult<T> = anyhow::Result<T, DynLoaderError>;

#[derive(thiserror::Error, Debug)]
pub enum DynLoaderError {
    #[error(transparent)]
    DataStore(#[from] store::StoreError),
    #[error(transparent)]
    Other(#[from] anyhow::Error),
}

impl DynLoader {
    pub fn new(store: Arc<Mutex<dyn store::Store>>) -> Self {
        Self { store }
    }

    pub async fn load_node_one(
        &self,
        id: &DynNodeId,
    ) -> DynLoaderResult<Option<DynNode>> {
        let store = self.store.lock().await;
        let object = store.obj_get(id).await?;
        Ok(object.map(|object| DynNode::new(id.clone(), object)))
    }
    pub async fn load_node_many(
        &self,
        ids: &[DynNodeId],
    ) -> DynLoaderResult<Vec<Option<DynNode>>> {
        // let ids = ids.iter().map(|id| id.to_string()).collect::<Vec<_>>();
        let store = self.store.lock().await;
        let objects = store
            .obj_get_many(&ids.iter().map(|id| id.to_string()).collect::<Vec<_>>())
            .await?;
        Ok(objects
            .into_iter()
            .enumerate()
            .map(|(i, object)| object.map(|object| DynNode::new(ids[i].clone(), object)))
            .collect())
    }
    pub async fn load_node_edge_count(
        &self,
        id: &DynNodeId,
        edge_type_id: &DynEdgeTypeId,
    ) -> DynLoaderResult<u64>
    where
        Self: 'static,
    {
        let store = self.store.lock().await;
        let key = store::AssocKey(id.to_string(), edge_type_id.to_string());
        let count = store.assoc_count(&key).await?;
        Ok(count)
    }
    pub async fn load_node_edges(
        &self,
        id: &DynNodeId,
        edge_type_id: &DynEdgeTypeId,
    ) -> DynLoaderResult<Vec<DynEdge>> {
        let store = self.store.lock().await;
        let key = store::AssocKey(id.to_string(), edge_type_id.to_string());
        let assocs = store.assoc_range(&key, 0, 9999).await?;
        Ok(assocs
            .into_iter()
            .map(|assoc| DynEdge::new(edge_type_id, assoc.id().clone(), assoc))
            .collect())
    }
}
