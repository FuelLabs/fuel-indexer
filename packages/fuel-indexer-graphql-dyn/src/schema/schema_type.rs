use super::data::*;
use super::edge::*;
use super::node::*;
use super::self_prelude::*;

#[derive(Clone, Debug, Serialize, Deserialize, Default)]
pub struct DynSchemaType {
    pub data: IndexMap<DynDataTypeId, DynDataType>,
    pub node: IndexMap<DynNodeTypeId, DynNodeType>,
    pub edge: IndexMap<DynEdgeTypeId, DynEdgeType>,
}

impl DynSchemaType {
    pub fn data(&self, id: &DynDataTypeId) -> &DynDataType {
        match id {
            DynDataTypeId::ID => Some(&DynDataType::ID),
            DynDataTypeId::Name(name) => self.data.values().find(|data_type| {
                dbg!(&data_type.name(), name);
                data_type.name() == *name
            }),
            _ => self.data.get(id),
        }
        .unwrap_or_else(|| {
            panic!("DynamicSchemaType::data_type: data type {:?} not found", id)
        })
    }
    pub fn node(&self, id: &DynNodeTypeId) -> &DynNodeType {
        self.node.get(id).unwrap_or_else(|| {
            panic!("DynamicSchemaType::node_type: node type {:?} not found", id)
        })
    }
    pub fn node_field(
        &self,
        id: &DynNodeTypeId,
        field_id: &DynDataFieldId,
    ) -> &DynNodeFieldType {
        self.node(id).fields.get(field_id).unwrap_or_else(|| {
            panic!(
                "DynamicSchemaType::node_field: field {:?} not found",
                field_id
            )
        })
    }
    pub fn edge(&self, id: &DynEdgeTypeId) -> &DynEdgeType {
        self.edge.get(id).unwrap_or_else(|| {
            panic!("DynamicSchemaType::edge_type: edge type {:?} not found", id)
        })
    }
    pub fn edge_field(
        &self,
        id: &DynEdgeTypeId,
        field_id: &DynDataFieldId,
    ) -> &DynDataFieldType {
        self.edge(id).fields.get(field_id).unwrap_or_else(|| {
            panic!(
                "DynamicSchemaType::edge_field: field {:?} not found",
                field_id
            )
        })
    }
}
