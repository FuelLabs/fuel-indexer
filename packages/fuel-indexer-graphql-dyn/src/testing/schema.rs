use super::schema_type::*;
use super::self_prelude::*;
use super::store::*;
use crate::schema::*;

use async_graphql::dynamic::*;
use graphql_parser::{parse_schema, Style};

#[extension_trait]
pub impl TestSchema for Schema {
    fn build_test() -> DynSchemaBuilder {
        let store = new_test_store().unwrap();
        let schema_type = new_test_schema_type(&store.r#type).unwrap();
        DynSchemaBuilder::new(&schema_type, Arc::new(Mutex::new(store)))
    }

    fn pretty_sdl(&self) -> String {
        let sdl = self.sdl();
        let ugly = parse_schema::<&str>(&sdl).unwrap();
        ugly.format(&Style::default())
    }
}

#[cfg(test)]
mod tests {
    use assert_matches::assert_matches;
    use insta::assert_snapshot;

    use super::*;

    #[tokio::test]
    async fn sdl() {
        // Build the schema
        let schema = Schema::build_test();
        let schema = schema.finish();
        assert_matches!(schema, Ok(_));
        let schema = schema.unwrap();

        // Print the schema
        let text = schema.pretty_sdl();
        assert_snapshot!(text);
    }
}
