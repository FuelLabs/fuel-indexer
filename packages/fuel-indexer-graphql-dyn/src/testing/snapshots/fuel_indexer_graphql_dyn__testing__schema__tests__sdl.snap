---
source: packages/fuel-indexer-graphql-dyn/src/testing/schema.rs
expression: text
---
scalar B256

type Block implements Node {
  id: ID!
  number: U64!
  hash: B256!
  parentHash: B256!
  chain: Chain!
  transactions(first: Int, after: String, last: Int, before: String): BlockHasTransactionConnection!
}

type BlockHasTransactionConnection {
  totalCount: Int!
  nodes: [Transaction!]!
  edges: [BlockHasTransactionEdge!]!
  pageInfo: PageInfo!
}

type BlockHasTransactionEdge {
  node: Transaction!
  cursor: String!
}

scalar Bytes

type CallReceiptData {
  param1: U64!
  param2: U64!
}

type Chain implements Node {
  id: ID!
  number: U64!
  blocks(first: Int, after: String, last: Int, before: String): ChainHasBlockConnection!
  transactions(first: Int, after: String, last: Int, before: String): ChainHasTransactionConnection!
}

type ChainHasBlockConnection {
  totalCount: Int!
  nodes: [Block!]!
  edges: [ChainHasBlockEdge!]!
  pageInfo: PageInfo!
}

type ChainHasBlockEdge {
  node: Block!
  cursor: String!
}

type ChainHasTransactionConnection {
  totalCount: Int!
  nodes: [Transaction!]!
  edges: [ChainHasTransactionEdge!]!
  pageInfo: PageInfo!
}

type ChainHasTransactionEdge {
  node: Transaction!
  cursor: String!
}

interface Node {
  id: ID!
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: String
  endCursor: String
}

type Query {
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]!
  chain(id: ID!): Chain
  block(id: ID!): Block
  transaction(id: ID!): Transaction
  receipt(id: ID!): Receipt
}

type Receipt implements Node {
  id: ID!
  type: ReceiptType!
  data: ReceiptData!
}

union ReceiptData = CallReceiptData | ReturnReceiptData

enum ReceiptType {
  CallReceipt
  ReturnReceipt
}

type ReturnReceiptData {
  val: U64!
}

type Transaction implements Node {
  id: ID!
  index: U64!
  hash: B256!
  gasLimit: U64!
  gasPrice: U64!
  blockHash: B256!
  chain: Chain!
  block: Block!
}

scalar U16

scalar U32

scalar U64

scalar U8

scalar Unit

schema {
  query: Query
}

